### DISCUSSION TITLE
Integration tests

### DISCUSSION THREAD

**Discussion by zc-devs** [woodpecker-ci/woodpecker#3262] [General]
@6543 [wrote](https://github.com/woodpecker-ci/woodpecker/issues/3178#issuecomment-1902484644):
> can we mock the kubernetis api somehow and create a integration test for the backend (with a smal pipeline as run example (1x clone 1x service 3xstep (2xparalel and one waiting))?

In case of #3178 mocking API doesn't make sense to me - looks like writing another implementation of Kubernetes.

It would be great, though, if we have intergration-integration tests (e2e). Test pipeline can be prepared in Gitea. Then in e2e test we can call WP API to trigger that pipeline. Then just check status of steps after some delay, maybe logs also. Instead of API, perhaps, CLI can be used.
Maybe some frameworks exist for this purpose.

PS: made discussion in order to not be lost.

**Comment by anbraten** [comment]
https://playwright.dev/ is pretty solid in terms of e2e testing UI, logic and components. For the test instance we can use the same setup we use for gitpod already: https://github.com/woodpecker-ci/woodpecker/tree/main/contrib/woodpecker-test-repo/.woodpecker. For a minimal kubernetes we could have a service running k3s.

**Comment by 6543** [comment]
there exist a dedicated dummy backend witch allos you to controle the step/workflow outcome

so we can be backend independent for creating an integration test for woodpecker ...
.. backend specific integration tests are also an idea but I would not focus on that one jet

https://github.com/woodpecker-ci/woodpecker/tree/main/pipeline/backend/dummy

**Reply by anbraten** [reply]
started some work some time ago: https://github.com/woodpecker-ci/woodpecker/pull/4954

**Comment by 6543** [comment]
# Create "backend integration test" framework -> #3819
